### ISSUE TITLE
Understanding `ChainRulesCore.rrule` in custom ReLU example

### ISSUE THREAD

**Issue by JinraeKim** [jump-dev/DiffOpt.jl#176]
Hi, developers! Again, it's me, Jinrae.

After understanding the ReLU example (#168), I'm trying to "use" DiffOpt.jl.
Unfortunately, I didn't get it properly.

If I understand the mechanism, it might be possible for me to add my understandings to the documentation (as a newbie).

# What I want to do using DiffOpt.jl
I want to make my custom neural network differentiable via Flux.jl (namely `my_NN`), which contains a convex optimisation problem.
For example, the loss function would be the mse error of
```julia
y - my_NN_optimal(x)
```
where `y` is the label, `x` is condition, and `my_NN_optimal(x) = my_NN(x, u_)` where `u_` is the minimised of `u -> my_NN(x, u)`.
To update `my_NN`'s parameters, I have to get some `rrule`.

# My understanding up to now
## What users can do using DiffOpt.jl
First, I've found in [Introduction](https://jump.dev/DiffOpt.jl/dev/intro/) that DiffOpt.jl can be used for differentiating the optimal solution to given parametrised convex optimisation problem with respect to problem parameters.
For example, let `x_` and `p` be the optimal solution and problem parameters, respectively.
It is possible to make an `rrule` for `dx_ / dp` using DiffOpt.jl.

## Strategy
As presented in some pages in docs, for example, [custom ReLU example](https://jump.dev/DiffOpt.jl/dev/examples/custom-relu/), I may have to add some `rrule`s (reverse-mode differentiation), which will be imported from [ChainRulesCore.jl](https://github.com/JuliaDiff/ChainRulesCore.jl).

## Where I felt stuck in
I'm struggling with the following things.
### What do `dx` and `dy` mean in [the custom ReLU example](https://jump.dev/DiffOpt.jl/dev/examples/custom-relu/)?
In the tutorial, especially where `ChainRulesCore.rrule` is defined, I cannot exactly follow what each line means.
It is probably due to my poor understanding of DiffOpt.jl and related packages, e.g., JuMP.

For example, this line means "performing optimisation" (with manually written function `matrix_relu`).
```julia
    pv = matrix_relu(y, model = model)
```
Then, what do the following lines mean? I mean,
- What is `dx`?
- I still don't understand even what `BackwardInVariablePrimal` is for.
```julia
            MOI.set.(
                model,
                DiffOpt.BackwardInVariablePrimal(),
                x,
                dx[:, i]
            ) # set sensitivities
```

Also, the following outputs `obj_exp` (seems an expression) but I don't get any clue about what it means.
```julia
            obj_exp = MOI.get(
                model,
                DiffOpt.BackwardOutObjective()
            ) # return grdiente wrt objective function parameters
```
For example, some outputs from REPL say
```julia
obj_exp = MOI.get(model, DiffOpt.BackwardOutObjective()) = -0.2500000001711336 x[1]² - 0.5 x[1]
-0.2500000001711336 x[1]² - 0.5 x[1]
```
so it is a quadratic expression of `x`. 
- What is `BackwardOutObjective` for?
- Why is it "the gradient wrt objective function parameters"?
- What's objective function parameters in this tutorial?

And some other questions:
- Which scheme do I have to choose among [two schemes](https://jump.dev/DiffOpt.jl/dev/manual/)?
- How can I readily transform my convex optimisation problem into a conic form (QP form would probably be restrictive in my case)?

Sorry for the verbose questions because of my lack of background for differentiable convex programming.
Thanks always.

**Comment by JinraeKim** [comment]
This is a modified code to observe each variable's value:

- code
```julia
using Test
using JuMP
using DiffOpt
# using SCS
using OSQP
import ChainRulesCore
using Flux


function matrix_relu(
    y::AbstractArray{T};
    # model = Model(() -> DiffOpt.diff_optimizer(SCS.Optimizer))
    model = Model(() -> DiffOpt.diff_optimizer(OSQP.Optimizer))
) where T
    _x = zeros(size(y))
    N = length(y[:, 1])
    empty!(model)
    set_silent(model)
    @variable(model, x[1:N] >= 0)
    for i in 1:size(y)[2]
        @objective(
            model,
            Min,
            x'x -2x'y[:, i]
        )
        optimize!(model)
        _x[:, i] = value.(x)
    end
    return _x
end

function func_test(
        y::AbstractArray{T};
        model = Model(() -> DiffOpt.diff_optimizer(OSQP.Optimizer)),
    ) where T
    pv = matrix_relu(y, model = model)
    @show model
    function pullback_matrix_relu(dx)
        x = model[:x]
        dy = zeros(T, size(dx))
        for i in 1:size(y)[2]
            MOI.set.(
                model,
                DiffOpt.BackwardInVariablePrimal(),
                x,
                dx[:, i]
            ) # set sensitivities
            DiffOpt.backward(model) # compute grad
            @show dx
            @show y
            @show obj_exp = MOI.get(
                model,
                DiffOpt.BackwardOutObjective()
            ) # return grdiente wrt objective function parameters
            @show dy[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in -2x'y
            @show dy[:, i] = -2 * dy[:, i]
        end
        return (ChainRulesCore.NoTangent(), dy,)
    end
    return pv, pullback_matrix_relu
end

@testset "diffopt" begin
    m = 1
    d = 1
    y = ones(m, d)
    @show matrix_relu(y)
    pv, pullback_matrix_relu = func_test(y)
    dx = ones(m, d)  # given x's value
    pullback_matrix_relu(dx)  # dy/dx
end
```

- result
```julia
matrix_relu(y) = [1.0000000006845344;;]
model = A JuMP Model
Minimization problem with:
Variable: 1
Objective function type: QuadExpr
`VariableRef`-in-`MathOptInterface.GreaterThan{Float64}`: 1 constraint
Model mode: AUTOMATIC
CachingOptimizer state: ATTACHED_OPTIMIZER
Solver name: OSQP
Names registered in the model: x
dx = [1.0;;]
y = [1.0;;]
obj_exp = MOI.get(model, DiffOpt.BackwardOutObjective()) = -0.2500000001711336 x[1]² - 0.5 x[1]
dy[:, i] = JuMP.coefficient.(obj_exp, x) = [-0.5]
dy[:, i] = -2 * dy[:, i] = [1.0]
Test Summary: |
diffopt       | No tests
Test.DefaultTestSet("diffopt", Any[], 0, false, false)

```

**Comment by JinraeKim** [comment]
After inspecting the source code and ChainRulesCore.jl's docs, I found that
`dx` denotes `dfoo/dx` for an output, namely `foo`.

Then, the output `dy` of the pullback should be calculated as `dfoo/dx * dx/dy`.
Still, I'm confused with what `obj_exp` means.
Why is it "a quadratic function in `x`"?

**Comment by JinraeKim** [comment]
I got it!
Since it was hard to me to fully understand the example without inspecting source codes and papers,
for those of you who will encounter similar issues (due to lack of understanding), I leave some comments.

# Preliminaries
I highly recommend you to read [Introduction, ChainRulesCore.jl](https://juliadiff.org/ChainRulesCore.jl/dev/index.html#Example-of-using-ChainRules-directly) to integrate such functionality with other AD packages.
Also, the introduction will be useful to understand the backpropagation mechanism.

## QP problems
For now, DiffOpt.jl provides mainly two cases: QP or conic problem.
I haven't tried conic cases, and the tutorial is about QP.
Consider the following QP problems as shown in [DiffOpt.jl's docs](https://jump.dev/DiffOpt.jl/dev/usage/).
![image](https://user-images.githubusercontent.com/43136096/145327339-02a188b7-b7ef-44a0-833b-545a40ea2d4a.png)

### Differentiate QP problems
In this regard, [Section 3: OptNet, Amos's thesis](file:///Users/jinrae/Downloads/bamos_thesis.pdf) is very helpful when inspecting the source codes.
Note that Amos takes slightly different notations as
![image](https://user-images.githubusercontent.com/43136096/145327625-fd8e5b24-0b8b-4b3f-af6e-e02c27867ef7.png)
and differentiation can be performed as
![image](https://user-images.githubusercontent.com/43136096/145327616-56b97ce2-ece4-4c01-ba7f-1dfb033f2f64.png)


## Back propagation
While performing backpropagation, it should be pointed out that most packages DO NOT calculate "pure" Jacobians.
For example, suppose backpropagation is started at the end with a loss function `l`.
And think about a layer whose input is `a` and output is `b`.
If we have `dl/db`, `dl/da` can be calculated as `dl/db * db/da`.
In this situation, `rrule` receives `dl/db` and produces `dl/da`, not `dl/db` (more precisely, the output of pullback is `dl/da`).

# Answers
1. `dx` is the input of pullback, provided from `ChainRulesCore.rrule`. That is, `dl/db`.
2. For users, just follow the way of setting `BackwardInVariable` and `BackwardOutObjective` in the tutorial.
3. The same as 2.
4. The commented out text, "return grdiente wrt objective function parameters" refers to a quadratic equation. I don't know why the developers decided to provide a quadratic equation, but anyway the quadratic and linear coefficients correspond to `dl/dQ` and `dl/dq, respectively.
5. Still, hard to answer this question for me
6. Still, hard to answer this question for me

## Note
The multiplication of `-2` was the most confusing part for me.
This is necessary because we want to get `dl/dy` (`dy` in the tutorial), not `dl/dq`.
That is, the linear coefficient of the objective function corresponds to `dl/dq`, so we have to multiply it with `dq/dy`.
Since `q = -2y`, `dl/dy = dl/dq * dq/dy = -2 * dl/dq`.
That's it!

**Comment by joaquimg** [comment]
Great that you figured all that.
Feel free to PR to the manual :)

**Comment by JinraeKim** [comment]
It might be possible after further understanding to this package,
for example, [#177, some comments](https://github.com/jump-dev/DiffOpt.jl/pull/177#pullrequestreview-827283036).
